### DISCUSSION TITLE
When do we need to use the Patient ID and when do we use a Patient Identifier ?

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#482] [Ideas]
When a user is created on the app, the  user is created using a UUID, while when the user is directly added to HAPI FHIR server via a POST command the user is created using an incremented ID 

1. Which criteria should we use?
2. When we want to retrieve a patient, do we retrieve /search using Patient ID or do we use the Patient Identifier?

**Comment by pld** [comment]
incremented IDs are preferably only for internal use since incremented IDs are less secure, so we probably use the UUID

**Comment by f-odhiambo** [comment]
Check out this usage here 

```
{
  "resourceType": "Patient",
  "id": "969", <= Let HAPI increment , do not submit as UUID
  "identifier": [ {
    "use": "official",
    "system": "http://who.int/ddcc/hcid", <= Can be https://fhir-web.opensrp-stage.smartregister.org/
    "value": "111000111" <= Add the UUID here 
  } ],
  "active": true,
  "name": [ {
    "family": "Murphy",
    "given": [ "Eddie" ]
  } ],
  "telecom": [ {
    "system": "phone",
    "value": "744585"
  } ],
  "gender": "male",
  "birthDate": "1986-09-19",
  "address": [ {
    "city": "Rome",
    "country": "Italy"
  } ]
}
```

**Reply by f-odhiambo** [reply]
This will definatey affect how we fetch and retrieve patients. Also need to check if this will apply to all resources

**Reply by pld** [reply]
yea definitely, since we'd fetch by identifier.

we should fact check this w/others in the FHIR community, if using the top level ID is what's normal, I'm ok doing that but then we'll need to make sure the ID assignments are the same local and on server, maybe it's fine to query by ID but never send an ID so HAPI is always assigning the IDs.

in current OpenSRP we assign IDs to a device before patients are created and then assign IDs from the device to the patients, but we have them reserved so the server doesn't give them out to others, not sure if we still need to do that, unsure if there are edge case problems we'd have if we don't do the batch pre-assignemnt to device.

CC @ndegwamartin for background on current opensrp

**Reply by pld** [reply]
Derek stated that from experience querying and joining based on identifier is not performant, not usable. Based on this he argued for using GUIDs as IDs everywhere (eg no sequential IDs), letting the client and server generate their own GUIDs without having to reserve any, and using the GUIDs as the reference IDs between resources
